### DISCUSSION TITLE
Return type of `data` when `placeholderData` is provided

### DISCUSSION THREAD

**Discussion by zachdtaylor** [TanStack/query#2278] [Q&A]
I noticed that even when `placeholderData` is provided as a `useQuery` option, the `data` attribute on the result can still be undefined according to TypeScript. Is that accurate? I thought the `data` attribute would never be undefined if `placeholderData` was provided. For example, if I do

```typescript
const { data } = useQuery<string>('key', () => fetch('/get-string'), { placeholderData: 'Loading...' })
```

the type of `data` is `string | undefined`, but I thought it should just be `string` or maybe `string | 'Loading...'`.

**Comment by TkDodo** [comment]
Yes, it’s a request that comes up a lot, but neither placeholderData nor initialData can currently narrow the type.
